### ISSUE TITLE
Improve the exception message of ToDictionary when keys are duplicated

### ISSUE THREAD

**Issue by yufeih** [dotnet/corefx#1187]
`Enumerable.ToDictionary` makes it easy to turn an array into a dictionary. But when the keys are duplicated in the source array, it just throws an `ArgumentException` that says **An entry with the same key already exists**. It is hard to know what keys are duplicated.

To improve the exception message without adding extra overhead, the exception message can be modified to **$"An entry with the key { key }  already exists"**.

`ToDictionary` internally calls `Dictionary<TKey, TValue>.Add()`, so the improvement made to Dictionary.Add benefits all code that calls the method.

**Comment by ellismg** [comment]
We've had this sort of feedback before (around Dictionary.Add()) and I think it's a great idea.  I sort of thought we had already done it, but I guess not?

/cc @davkean

**Comment by pdelvo** [comment]
Problems I see with that:
- I think it is very likely that there are people who are using the message and comparing it to figure out what kind of ArgumentException it was.
- Now .ToString whould have to be called for an element. This could introduce side effects.
- .ToString maybe even throw another exception. What would you do then?

So I don't think it is worth the risk of breaking peoples code.

I have a different idea:

Create a new exception that inherits ArgumentException which has a property "DuplicatedItem" that contains the item that caused the exception to happen. This would solve all three problems I mentioned, but it could break compatability (.GetType() on the exception would return a different type now), but I think thats far more unlikely

**Comment by ellismg** [comment]
Related (and what I was thinking of: #185).

I don't think the message comparison argument holds water.  We localize the text of these messages anyway so folks are already on thin ice if they do this sort of comparison.  I think that we are okay to introduce breaks in these cases.

The "ToString()" having side effects or throwing exceptions angle is interesting, but we've [already said in the contract that you should do either of these things](https://msdn.microsoft.com/en-us/library/system.object.tostring%28v=vs.110%29.aspx).  Of course, folks could break the rules and maybe they get burned here, but again, I think the benefit is worth the reward. 

I actually think that throwing a subclass is more likely to cause issues (due to folks doing type tests and getting things wrong) than changing the exception text, even though we do allow throwing new derived exception types without considering it a breaking change.  It also forces folks to opt in to the correct behavior (catch this type, and print a helpful message (I assume that you would not call the duplicate element's ToString() in this new Exception's ToString() due to the same argument about not knowing what happens when ToString is called on some arbitrary object).

**Comment by yufeih** [comment]
So if this change makes into immutable collections, it should also be included in `System.Collections.Generic.Dictionary`. I am not sure if the non generic `Hashtable` needs this, as that class is considered obsolete.

I can help come up with a pull request if this change Is considered acceptable.

**Comment by akoeplinger** [comment]
> So if this change makes into immutable collections [..]

@yufeih It already did, see https://github.com/dotnet/corefx/pull/185.

I'm just waiting for `Dictionary` to show up in corefx to send a similar PR :smile: Can't wait to have this finally fixed :+1:

**Comment by davkean** [comment]
Dictionary is coming soon.

**Comment by mellinoe** [comment]
Yes, Dictionary should be coming soon, hopefully within the next two weeks.

**Comment by akoeplinger** [comment]
@ellismg is there still something missing in CoreFx to do this? Or should this be done in CoreCLR given that `Dictionary<TKey,TValue>` lives there?

**Comment by ellismg** [comment]
Talking with @mellinoe it sounds like we still think we could pull `Dictionary<TKey,TValue>` out of coreclr and into corefx, but there's no one signed up to do the work (and I think the priority of pulling it out is pretty low when compared to the other work on our plate)  I think we should just go ahead and do the work in coreclr today and then if we do pull it out we'll port the fix over as part of that work.

**Comment by akoeplinger** [comment]
I took a stab at this in https://github.com/dotnet/coreclr/issues/1452.

**Comment by OtherCrashOverride** [comment]
All these 'better' exception messages need to be opt-in or at least opt-out.  They present lots of issues for little to no benefit.

1) It leaks potentially sensitive information.  Its not uncommon for an ASP.Net app to display the exception on a web page.
2) It open the possibility for HTML/SQL injection when output to log files or web pages.
3) It can be used as a denial of service method.  A string can be 2^32 characters (UTF-16).  Writing this to a log can easily exhaust it.  On devices with slow storage (mobile) its even more of an issue.
4) It breaks localization.  The error message may be En.US-UTF16, but the key may be some other language.
5) Since there is no localization guarantee, its possible the string may be visually non-sense on the log and/or terminal.
6) UTF string comparison rules may not be what the developer expects.

Its enough to know that you have the exception.  If a developer requires more information, they can catch the exception and log additional information taking the precautions necessary.  Exception messages should not be a replacement for a debugger (printf debugging).  A proper debugger and core dump feature is the solution here.

**Comment by MgSam** [comment]
@OtherCrashOverride As mentioned in the beginning of this thread, the problem in Dictionary is that you cannot figure out what the problem was by catching and logging the exception, especially in the context of Enumerable.ToDictionary. You are forced to write additional code to do analysis or step through a debugger. The potential added benefit for after-the-fact debugging that would result from improving the exception is huge.

I do agree though with your list that this change has potential security and localization implications. 

It seems to me a better way of implementing this would be added a `Value` property to ArgumentException that contains the offending value. This would then not break any existing consumers depending on the existing message and would not introduce any of the string-related concerns mentioned.

**Comment by jasonwilliams200OK** [comment]
> \*security concerns\*

debatable..

Does any other language (nearly as *cool* as C#) show the keys? If they do, then C# should too.

> \* existing code \*

If the dev is re-targeting new core anyway, they should read the docs and keep this in mind that they should handle the dictionary keys exceptions "if they have \*super sensitive\* information && they cannot come up with a less obvious name for a key && they got D in cryptography".

BTW injection attack can be made on dictionary keys itself! In a broader context, there are more mechanisms to prevent from such repetitive throttling access situations and the chances of Dictionary object getting hijacked are as good as: *everything is hackable*.

**Comment by OtherCrashOverride** [comment]
We should probably get rid of Enumerable.ToDictionary().  It is obviously a poor design choice.  There is no reasonable expectation that anything enumerable should be convertible to a dictionary.  Where the assumption that it can be turns out to be false is where the issue comes to light since there is no way to report why it failed.  Its also not reasonably expected that it should report why. The enumeration can be an arbitrarily large ranging in terabytes or zetabytes (enumerations have no limit).  The enumeration can also contain an arbitrary amount of duplicate keys.  Knowing what key is duplicated is not of any benefit.

A better solution would be to add ``bool Dictionary.TryAdd(K key, T value)``.  The developer can then report information associated with failures that makes sense.

**Comment by OtherCrashOverride** [comment]
It should also be noted that reporting the key as string in an error message has only limited benefit to those using human readable information as a dictionary key: this means strings and numbers.  For the cases where a key is an object and ToString() reports the object name, its of absolutely no benefit.

**Comment by JonHanna** [comment]
> There is no reasonable expectation that anything enumerable should be convertible to a dictionary.

It's a perfectly sensible expectation, especially since the caller defines the terms by which the keys and values are defined.

> It should also be noted that reporting the key as string in an error message has only limited benefit to those using human readable information as a dictionary key.

This I agree with. Subclassing the exception with one that provides access to the key would be more useful here, as well as not turning current code that doesn't leak information into code that does.

**Comment by OtherCrashOverride** [comment]
>There is no reasonable expectation that anything enumerable should be convertible to a dictionary.

This means that what ever does the conversion SHOULD expect errors.  There is no reasonable expectation that the process will be successful. This is why its a bad design.  Adding the TryAdd method allows developers to write extension methods or use lambdas to perform the operation.  Additionally, it gives them control over what happens when the process fails.  If you have a million entries with 25,000 duplicate keys, logging or reporting ONE of them is pointless.  With TryAdd, the developer can log the duplicates in some meaningful way or even ignore them and proceed without terminating the application.

**Comment by akoeplinger** [comment]
@OtherCrashOverride 

> 1) It leaks potentially sensitive information. Its not uncommon for an ASP.Net app to display the exception on a web page.

Showing exceptions to users has long been strongly discouraged. It is off by default for this reason.

> 2) It open the possibility for HTML/SQL injection when output to log files or web pages.

If you're accepting user input, you're responsible for sanitizing it. Nothing changed in this regard.

> 3) It can be used as a denial of service method. A string can be 2^32 characters (UTF-16). Writing this to a log can easily exhaust it. On devices with slow storage (mobile) its even more of an issue.

Again, you're responsible for sanitizing user input. Putting a multi-gigabyte string as a key in a dictionary is going to exhaust your RAM even before you write it to a logfile.

> 4) It breaks localization. The error message may be En.US-UTF16, but the key may be some other language.

I don't see how this "breaks" localization anymore than getting "Error HRESULT 0x803242" or other common messages does.

> 5) Since there is no localization guarantee, its possible the string may be visually non-sense on the log and/or terminal.

Yeah, that's a valid point. I still think it's better to have readable output most of the time (you're in control of what to use as a key) than nothing at all.

> 6) UTF string comparison rules may not be what the developer expects.

How is this relevant here? Do you mean when a developer compares exception message strings?

Summing up: none of your points convince me this shouldn't be done. In addition, the maintainers already decided in https://github.com/dotnet/corefx/pull/185 that leaking information to exception messages is OK :)

**Comment by masonwheeler** [comment]
`We should probably get rid of Enumerable.ToDictionary(). It is obviously a poor design choice. There is no reasonable expectation that anything enumerable should be convertible to a dictionary. `

[citation needed]

`var index = data.ToDictionary(e => e.Name);`  That's (a slightly anonymized version of) something I've actually done in real-world code.  What's unreasonable about it?

**Comment by OtherCrashOverride** [comment]
> In addition, the maintainers already decided in #185 that leaking information to exception messages is OK :)

Yes, it has become clear that this project is in desperate need of a fork long before it has even been released.  As always, I hoped there would be some kind of compromise that suited everyone.

**Comment by terrajobst** [comment]
> Yes, it has become clear that this project is in desperate need of a fork long before it has even been released. As always, I hoped there would be some kind of compromise that suited everyone.

Wow, that's quite a jump. I think @akoeplinger did a good job of responding to your concerns. You conclusion essentially reads as "I don't agree so I need to fork". First of all, I don't want to dismiss your arguments but I also feel the burden is on you to convince us of them. That's the point of a technical debate. It's based on merit and that's based on arguments.

Secondly, if you should still feel we're doing a bad job of maintaining the .NET Core framework then you're absolutely entitled to fork it. The nice thing about open source is that you don't have to convince the owners -- you can always unblock yourself and control your own destiny by creating a fork.

However, I'm not sure that's making your cause any easier. Unless you only care about having your own copy of the framework you'd have to convince the community that your fork is indeed a better version. So while you don't have to convince us, you certainly have to convince somebody.

So either case I think we're all better of hearing what you've to say, so let's hear it!

**Comment by migueldeicaza** [comment]
The road to CVEs is paved with people that considered that security concerns over leaked information were "debatable".

**Comment by akoeplinger** [comment]
I appreciate the discussion around the possible implications, it's great that we can debate the pros/cons of such changes here even if not everyone agrees. 

In the end the maintainers need to decide whether it's a good change and I think the root question is whether keys in a dictionary typically hold sensitive information (personally I'd argue you're doing something wrong if they do). In any case, please keep the feedback coming :smile:

**Comment by terrajobst** [comment]
That's fair but I don't quite know how that is actionable because it applies to any information contained in the exception is subject to potential exposure. This includes, the type, the message, and obviously any value from the program, which is what's being discussed here.

I'm with @akoeplinger that the only sane way to address this in systems is by making sure that exceptions aren't leaked.

**Comment by advancedrei** [comment]
In an era of RUM systems like New Relic and Application Insights, I'm kind of shocked this is even a debatable topic. Though I guess if you get a scientist and a philosopher in a room, they could debate if the sky was blue.

Exception messages are supposed to help the developer. The exception message here is terrible. It doesn't help the developer understand what happened *at all*. Full stop. 

Exceptions are not supposed to be displayed to the end user. If they are, that's on the developer, not the Framework. It is not the Framework's responsibility to save me from myself. Full stop.

There are terrible exception messages all over the Framework. They should be fixed. in the time this thread has been debated, an intern could have fixed them all and moved on to less obvious problems.

Sorry if this comes across as rude or harsh, but both of the threads discussing this are kind of ridiculous. There is such a thing as overthinking something.

**Comment by migueldeicaza** [comment]
sensitive information is not limited to the login and password for your web site, but it includes any information than an attacker can use to his advantage.

This is just generally accepted by those in security circles, and specially those that had an egg in their face after information leakage lead to giving attackers information.

I can make up hundreds of scenarios of information that i could use to gather information about a system to hack it, and none have to include your login or password.

This is just a well known problem in the space, not even sure why this is even remotely debatable

**Comment by davkean** [comment]
Miguel, I think the argument boils down to; it is up to the BCL to protect against information leakage and/or protected against malicious input (think XSS)? Or is this better left to the app models themselves? 

Path information could also be considered leakage - so the same argument would apply to any IOException that happened to include paths, plus the many other exceptions that include input that might come from the user or contain information about a system.

**Comment by JoshVarty** [comment]
@migueldeicaza As someone who builds applications for the desktop (where this wouldn't be a security issue), I'm not sure if it makes sense to make my life harder because some web developers have explicitly ignored a default and decided to display exceptions to their users.

**Comment by davkean** [comment]
One more thing, I remember from a few attacks we've had in the past, the very fact that a FileNotFoundException was thrown or the call stack itself (!) could give hints or information about a system, regardless of message.

**Comment by advancedrei** [comment]
The whole stack trace can give attackers more than enough to do damage. Common application hardening practices say to shut off the error page. It's not the framework's job to do that. As a developer, especially if I'm reporting exceptions to a central exception logger, I need to know exactly what happened. And the exception message should give me that. If I leaked the exception to the user, and I get hacked, that's my fault. To answer @davkean's question, is not the responsibility of the core framework to protect against information leakage. That is the job of ASP.NET or WinForms or WPF or whatever higher level app framework.

This exception is the equivalent of your car throwing a yellow engine light and expecting you to hook up a diagnostic machine whenever there is a problem. Cars nowadays are often smart enough to be able to tell you the problem without a ODBII debugger attached.

**Comment by masonwheeler** [comment]
@JoshVarty has the right of it.  This is disabled by default on ASP.NET.  For it to become a security concern you literally have to go out of your way to *ask to be hacked*.  And on client-side software it's not a security concern, so why worry about it?

The simple fact is, the only people who are going to be burned by this are people who deliberately engage in acts of willful stupidity, and there's really no use trying to help *them*.  Try to engineer something to make it idiot-proof, and the Universe will out-engineer you by building a better idiot.

**Comment by agocke** [comment]
@migueldeicaza I'm not exactly sure what position you're arguing for. I think the argument you're making, which is the same one I would make, is that leaking any information, at all, about internal errors is a security risk. No information, no matter how innocuous, is safe to show a user. It's not even safe to tell them what kind of error was encountered, only that an error was encountered.

If that is correct, then there should never be an instance in which an exception should be shown to the user. The only purpose of an exception message is to help the developer debug their application, in which case the message should be as useful as possible.

This line of reasoning means we should never even consider the security implications of an exception message, because their presence in a security context is inherently a massive security vulnerability. Do you agree with this categorization?

**Comment by marklio** [comment]
The exception is not normally bad because it's thrown at the callsite where you have the value. ToDictionary obscures the callsite, but it's implementation is trivial. Write your own that has an exception callback, or that has whatever semantics you want. Don't worry too much about complicating compatibility or security.

**Comment by Vannevelj** [comment]
What would be the downsides of a compiler directive to only expose the information in a debug environment? I believe that covers both arguments, though it will become rather messy if you have to do this everywhere.

**Comment by davkean** [comment]
@Vannevelj do you have a design/feature in mind that would cover this?

**Comment by ericholscher** [comment]
If you generally believe error messages are a security concern, then a program should never show any useful error. This seems like a backwards way to think about building software. Developers need as much information as possible during development, but that should obviously not be exposed to end users in a production environment.

I think that @Vannevelj's point is a good one. Most of my experience is with Django, which has a DEBUG setting: https://docs.djangoproject.com/en/1.8/ref/settings/#debug -- it provides completely insecure information to the developer during development, but would never be used in a production setting. These are two totally different environments, and should behave differently.

I generally agree that error messages should be as useful as possible, and it's the responsibility of folks to not have them turned on or published in a production environment.

**Comment by jasonwilliams200OK** [comment]
Regarding ASP.NET applications, in addition to what @akoeplinger said about ASP.NET `customErrors` being turned off by default, CoreFX is probably *only* going to entertain ASP.NET5+ applications via DNX, right? Then this blog is related: http://weblogs.asp.net/imranbaloch/custom-errors-aspnet5-mvc6. The `customErrors` in `Web.config` is gone in favor of middleware opt-in/opt-out approach for error pages with exception message (or a fancy one with unicorn sticker saying "What is the difference between unicorn the page you are looking for? they both don't exist!"). Which means now we need to explicitly "configure" an error page to leak the exception information to the mainstream audience. For production, the exception details can be channeled from middleware to RUM systems directly (as @advancedrei indicated), which is also very possible in ASP.NET4 and lower versions.

@migueldeicaza is right that concealing "keys" or labelled "security" details do not suffice to mitigate chances of getting hacked; a capable hacker can work with a mere call stack to discover potential loopholes. 

IMO, if majority of devs are agreeing in this thread, here: https://github.com/dotnet/coreclr/pull/1452 and here: https://github.com/dotnet/coreclr/issues/1311 that exception messages should be transparent, then the few folks in opposition, instead of start trolling they should bring a strong argument to convince us or show some decency to submit to the decision being made based on the open argument..

To the contributors, collaborators, dotnet team and everyone who is sharing positive energy (via constructive criticism), thank you and keep up the great work!

**Comment by OtherCrashOverride** [comment]
> So either case I think we're all better off hearing what you've to say, so let's hear it!

@terrajobst for completeness I will re-iterate my suggestions here.

1) Add #ifdef feature tags around the 'enhanced' messages.  This will allow for easy custom builds that maintain the desktop .Net secure behavior.

or

2) Add a new field to ``System.Exception``.  ``Exception.Message`` can remain as-is and anyone wishing to opt into less secure behavior can use a new ``Exception.ExtendedMessage``.

I think for ``Exception.Message`` a basic rule of "It may only use interned strings" is a good criteria.

**Comment by OtherCrashOverride** [comment]
When an exception occurs, the code is in an exceptional state.  Its impossible to anticipate what side-effects calling ``ToString()`` on an object that was a cause of the exceptional state will have.  At the very least, I think this proposal needs some real world testing and documented experience.  As I sated before, this particular item provides no benefit to the majority of developers in the majority of circumstances.  However, it potentially negatively impacts all of them.

**Comment by OtherCrashOverride** [comment]
To show my pioneering spirit and forward thinking,  I will make this proposal now:

We should add facebook and twitter integration to System.Exception!

:+1: :+1: :100:

**Comment by zabulus** [comment]
Why this exception is so special? There are already exceptions in system.io that exposes internal applicatio data. "Could not find part of the path {0}" for example.

**Comment by masonwheeler** [comment]
> Why this exception is so special?

It's not.  The "security" argument is completely baseless and ridiculous, for reasons that have already been explained abundantly in this thread.

**Comment by OtherCrashOverride** [comment]
> There are already exceptions in system.io that exposes internal applicatio data.

Thank you for pointing that out. I will submit a pull request to remove it.

**Comment by yufeih** [comment]
The work to add **key** to the exception message is done, so I am closing this thread. 

Thanks everybody for driving this to reality! 

Please continue the debate here: https://github.com/dotnet/coreclr/issues/1459

**Comment by blowdart** [comment]
As @davkean  and @terrajobst  asked me to chime in last night whilst I was rebuilding my machine I'm ignoring closed...

As the keeper of ASP.NET security I view it as our (ASP.NET's) responsibility to sanitize exception messages from a core framework. We've changed our error handling so it's a plain 500 with no details by default, as the default environment is release, and you can only change that by environment variables, not by accidentally deploying the wrong config file.

In this specific case I don't have any real security concerns.

**Comment by OtherCrashOverride** [comment]
> In this specific case I don't have any real security concerns.

With all the certainty that this poses no security risks, perhaps Microsoft or The Dot Net Foundation can back up that assertion with a financial guarantee.

I have no doubt that ASP.Net will not expose exception information without user intervention.  Now everyone else in the world using the CoreCLR framework has to conduct code audits to ensure the same.

Yes, these changes will cost companies both time and money with no benefit to them.  These exception message changes are pointless for professional developers.  Having ONE duplicate out of MILLIONS in a log file is not going to change anyone's developer experience in any substantial or meaningful way.  If you are working on a HelloWorld program with 3 dictionary entries all using strings and just learning how to code, then there may be some potential benefit.

So what is the issue with NOT reporting these messages as ``Exception.Message`` and instead using a different property?  Is the potential benefit not worth the time to change one line in your code to get a more verbose message?  What is wrong with using a different #ifdef feature flag for them?  Is the potential benefit not worth changing a single dependency line in a DNX project file?

As long as CoreCLR is only seen as part of ASP.Net,  these poor decisions will continue to hamper the codebase.  How about this completely alternate proposal: make it work first, then worry about 'enhancing' it.

While this thread (and others) may have started as an intelligent discussion or debate.  They always digress into name calling and strawman arguments.  Scroll back through this thread and its easy to see why myself and others have lost patience.  At what point does it all become more trouble than its worth?

#DontWIndows8DotNetCore

**Comment by OtherCrashOverride** [comment]
For clarity I want to explicitly state: this message not only leaks the key, it leaks any and all data in an Object's ToString() all of which have NOTHING to do with the failure since Objects in a dictionary are reference equality not value equality.

**Comment by robertmuehsig** [comment]
I really don't see the security issue here, because it is not in the hand of the CoreFX to generate "safe" or "unsafe" error message and there are other exception message with "context" information - so I like this idea.

**Comment by OtherCrashOverride** [comment]
> I really don't see the security issue here

CoreCLR will now expose arbitrary information [ToString()] to logging mechanisms without the developers consent that may be viewed by person(s) not authorized to it.  The logs may be stored on media or locations where the information is not allowed due to policy or legal regulation.

I guarantee you that even Microsoft will now have to institute policy banning CoreCLR's use in certain areas.

**Comment by masonwheeler** [comment]
You keep making vague allusions to attackers, but never once do you specify any actual threat model.

If the client is running the software, they already have full access to it, and therefore there is no security issue.  If the client is not running the software, then the server can simply choose not to display exception messages to the end-user. (Which is the default on ASP.NET and you have to go well out of your way to change it, and on other server software it should either be the default or very easy to change.)

So since no threat exists on the server side, (unless the admin is a moron, which you really can't do anything about,) and no threat exists on the client side, *where does the threat exist?*  If you want your claim to be taken seriously, demonstrate a real-world use case where this could actually be an attack vector.

**Comment by OtherCrashOverride** [comment]
>  If you want your claim to be taken seriously

I'm not sure you appreciate how insignificant of a concern that is.  What you choose to believe or not has no correlation on what is or isn't real.  Its simply a matter of the due diligence in being a good internet citizen to point out the issue.  #You'reWelcome!

**Comment by jasonwilliams200OK** [comment]
#Don'tFeedTheTroll

**Comment by robertmuehsig** [comment]
@OtherCrashOverride Ok - I get your point, but there are currently other Exception Messages out there with context information, e.g. IOExceptions with path. This should be fixed as well, right?

Your point is: Don't write any "detailed" information inside Exception.Message, because in most cases this will be logged and stored inside unsafe logfiles.
Your solution would be: Keep the Exception.Message property clean and use Exception.DetailedMessage (or something like that) to store potential real context information and this DetailedMessage wouldn't be touched by "ToString" and therefor it is in the hand of the actual developer if he wants to commit those detailed information to the logging infrastructure.

Mhhh... I think this could be handy, but then each Exception.Message should be clean.

**Comment by terrajobst** [comment]
> With all the certainty that this poses no security risks, perhaps Microsoft or The Dot Net Foundation can back up that assertion with a financial guarantee.

Any bug can potentially become a security issue. Software is complicated and is virtually guaranteed to have bugs. Nobody is stating otherwise. But for practicality of day to day fixes and improvements it should be OK to make reasonable assumptions.

> I have no doubt that ASP.Net will not expose exception information without user intervention. Now everyone else in the world using the CoreCLR framework has to conduct code audits to ensure the same.

If "everyone else" implies a similar security sensitive thing like a web server, then the answer is. However, I highly doubt that's "everyone". For example, console and UI applications usually don't fall under this.

> Yes, these changes will cost companies both time and money with no benefit to them. 

I beg to differ. I've been seen numerous occasions with `KeyNotFoundException` or `ArgumentException` with not enough context to figure out what's going on. Sure, once you attach a debugger and repro the bug, it's not super hard to figure it out: but that does cost time. Same is true for error codes. Of course, it's easy to map Win32 error codes and HRESULT values to error messages but it sure is a heck of a lot more productive if the exception's error message contains the readable string already.

> So what is the issue with NOT reporting these messages as Exception.Message and instead using a different property? Is the potential benefit not worth the time to change one line in your code to get a more verbose message? What is wrong with using a different #ifdef feature flag for them? Is the potential benefit not worth changing a single dependency line in a DNX project file?

This was the kind of response I was hoping to get, by the way. Constructive feedback helps to move the discussion forward.

So let's see:

* Different property. Doable, but might be weird considering that the exception for duplicate keys is `ArgumentException`. We could, however, create a derivative of it (`DuplicateKeyException`) and add the property there. However, this doesn't necessarily address the underlying issue because unless the code handling exceptions serializes it to something readable, I still have to be in the debugger to help diagnose the issue.

* `#if`. What code would you guard with this? If it is the code that constructs the error message then it requires callers to compile their own version of the framework with additional diagnostics to get the benefits. I don't think this is practical.

Somebody on Twitter suggested a global flag that developers can turn on, such as `Environment.EnhancedExceptionMessages = true;`. I believe this is similar to what you suggested via `#if` but it has the benefit that it doesn't require recompilation. It would be up to the app model to decide whether better exception messages are worth the potential increase in exposure.

Thoughts on such an approach?

**Comment by OtherCrashOverride** [comment]
As touched on earlier in this thread, when these verbose message are of the most benefit is when you are actively developing.  Its worth exploring additional ideas where a debug environment provides very detailed information but a release build (of the runtime) does not.  Once a product has matured, the need for such exacting detail is diminished.  I suspect that once Visual Studio remote debugging is enabled and feature mature, the need for expanded exception messages will diminish.  Debugging is the problem that is trying to be solved here.  Better debugging tools should go a long way to alleviating any concerns about less concise exception messages. 

A global configuration flag presents its own issues.  It would need callsite guarantees that its not being enabled by 3rd party libraries.

It would probably be beneficial to hold all changes related to the debugging experience until we have a complete picture of what tools will offer.  Once we are feature complete in that area, then we can address any actual deficiencies rather than perceived ones.  A feature rich remote debugger makes a huge difference during product development.

**Comment by nosami** [comment]
@OtherCrashOverride I agree with you on the potential information leak issue here. I don't think better debugging tools is the answer though. I would rather not have to use a debugger at all.

**Comment by akoeplinger** [comment]
@OtherCrashOverride 

> Its worth exploring additional ideas where a debug environment provides very detailed information but a release build (of the runtime) does not. Once a product has matured, the need for such exacting detail is diminished. I suspect that once Visual Studio remote debugging is enabled and feature mature, the need for expanded exception messages will diminish.

I disagree with this. Debugging is **not** a replacement for logging in production, there are often times where you can't reproduce the issue while debugging and the only information you have are the logs or you don't have the ability to attach a debugger, e.g. because you don't control the system where your code is running. I recommend reading @ayende's great series of [Production Postmortem](http://ayende.com/blog/posts/series/171074/production-postmortem) blog posts for the sort of issues where debugging isn't sufficient and only extensive logging can help you diagnose.

I agree that #if for the detailed logging isn't a good solution as deploying a debug version of the framework/runtime isn't viable in these scenarios.

Having a flag to control the details (`--verbose` ?) would be a compromise that might be OK, but we'd need to go through the framework and scrub all cases like FileNotFoundException as well.

Personally, I'm still missing a good rationale why this really changes anything related to information leakage in exception messages, the core framework never made any guarantees in this regard (as e.g. FileNotFoundException and many other exceptions show). If we want to change the framework to make these sort of guarantees, then that's IMHO a different discussion.

**Comment by vbfox** [comment]
> As touched on earlier in this thread, when these verbose message are of the most benefit is when you are actively developing.

In the two situations I know best, my current employer (Client/Server apps for the financial market) and my previous one (Instrumentation apps for R&D or factory line installation, not connected to the internet) having more information available in exceptions would have been a great help.

In both cases while getting logs is possible with legal hurdles in one case and technical/human ones in the other) but reproduction or debugger attaching is often very hard.

Financial institutions have some very specific system installs (App-V or Citrix + a strangely behaved proxy are the norm) and installing anything new on the machines can be a month long marathon between services.
Factories often stop only a few hours per month and have hardware too costly to get, and R&D labs don't want anyone near their prototypes that nobody knows about.

The consequence is that often the only resource you have is often a long-cold log of what action happened and what exceptions were produced.

And in both cases nobody cares about the security of the exception messages in particular, they aren't treated separately from logs and logs contains things that are a lot more precious to our clients than exceptions could ever contain.

Note that having the information in some `FullMessage` property would be ok for both cases, the serializer for exception being the logging system it's a simple change to make it use this property as long as it is on a common interface `IExceptionWithFullMessage` that can be easily detected.

**Comment by OtherCrashOverride** [comment]
> Debugging is not a replacement for logging in production

I am not proposing that it is.  The developer should be in control of what is logged, not the runtime via exception messages.  There are numerous logging frameworks available.  Developers are free to add as much or as little logging information as desired.  The developer is the one in the best position to decide that, not the CoreCLR framework.

> I'm still missing a good rationale why this really changes anything related to information leakage in exception messages

ToString() greatly increases the amount of information unintentionally leaked when its overloaded and provides useless information (object type name) when its not.  More importantly, as explained earlier, for objects (not valuetypes) as keys, this information does even have a relation to the exception.  What the object contains is not relevant. Two different objects as keys could contain identical information; however, they are unique keys.

**Comment by masonwheeler** [comment]
> As touched on earlier in this thread, when these verbose message are of the most benefit is when you are actively developing.

What are you talking about?  When you're actively developing is when verbose error messages are the *least* useful, because when you're actively debugging you have the debugger right there for you.  The only time messages like this are useful is *in production,* because that's when you don't have the debugger, so they provide helpful information to reproduce the bug.

**Comment by OtherCrashOverride** [comment]
So to summarize the point we are at:

'Enhanced' exception message: some people don't like the idea.
Better remote debugging: some people don't like the idea.
Core dumps: some people don't like the idea.
Leave everything as it was: some people don't like the idea

**Comment by akoeplinger** [comment]
> The developer should be in control of what is logged, not the runtime via exception messages.

I agree to a degree, and I could certainly log the key before writing to the dict in this case. In other cases this isn't possible. Consider `Directory.GetDirectories(@"C:\Windows\system32", "*.*", SearchOption.AllDirectories);`. Right now, this throws UnauthorizedAccessException ("Access to the path C:\Windows\system32\Com\dmp\ is denied"). How would a developer know that beforehand and log it?

> ToString() greatly increases the amount of information unintentionally leaked when its overloaded and provides useless information (object type name) when its not.

That's a valid concern, if the ToString() doesn't provide uniquely identifying information then yeah it won't help me much. Would you think it's less of a problem if we constrained including the key to valuetypes like `string`, `int`?

**Comment by OtherCrashOverride** [comment]
> How would a developer know that beforehand and log it?

```
Log: Retrieving directories for [encrypted string]
Log: Caught UnauthroizedAccessException: exception.DetailedInfo=[another encrypted string since its a property of the exception and can be reported here in the log but is not written to the clear text log file bet you wish exception could do this automatically, hey maybe we should provide encryption functions then everyone would be cool with logging arbitrary information because its automatically encrypted by the runtime. there is a good idea to consider]
Log: Finished.
```

**Comment by OtherCrashOverride** [comment]
So for those that missed it in the above: maybe adding encryption support to exception message/detailedmessage may be an option

**Comment by blowdart** [comment]
Re: Encryption. Oh heck no. Now you have key management to consider. Too much effort for very little gain.

Frankly IMO if you're putting secrets in your ToString() you're doing something wrong, but that's just me.

**Comment by akoeplinger** [comment]
Yeah, key management is a major hassle. @OtherCrashOverride it's an interesting idea, maybe we need a way to tag certain pieces in an exception message as "sensitive" and something in the logging pipeline turns that into ******* when you have a certain flag set (or the other way round)?

**Comment by OtherCrashOverride** [comment]
Lets give it some time and see what others have to suggest.

**Comment by OtherCrashOverride** [comment]
Writing this down here for future discussion while its still fresh in my mind:

The solution I am leaning towards is:
Exception.Message and Exception.DetailedInfo - Message is reworked to use only strings interned (no dynamic messages).  DetailedInfo - is only activated when read.  If it is never read, code for it never runs. resources are never consumed for it. and there are no sideeffects of its use.  DetaledInfo provides as much information that makes sense in the context of the exception while avoiding over stating information.

DetailedExceptionInfoFeature flag that is passed to the runtime during creation by the host.  How the host determines to enable it (environment variable, command line option, registry entry) is entirely host dependent.  The flag is part of the hosting API and defaults to disabled if not explicitly enabled.  It can only be specified during host creation and is immutable.

When DetailedExceptionInfoFeature is not enabled Exception.DetailedInfo returns an interned string of "Feature Not Enabled".  No additional code is run, no additional resources are consumed and no side effects are present.

This models gives each developer the freedom to do what they will with exception messages at a level that is appropriate and optionally the ability to log, encrypt or discard the information.  It provides a single runtime binary encompasing the feature as its expected to not contribute a substantial amount of additional code if unused and will not consume any resources if not used.

**Comment by masonwheeler** [comment]
You're proposing to introduce a ton of additional complexity for no reason.  First off, how would you even enforce a rule like "no non-interned strings?"  Call `IsInterned()` on the string passed into the `Exception` constructor?  What happens if it returns `false`?  (The appropriate way to deal with invalid input is to throw an exception...)  This proposed change would break tons of existing code, and it provides no benefit as you *still have not demonstrated that an actual security threat model exists.*  Just about everyone else who looks at this, *including the guys in charge of the project, who really ought to know,* have said that it does not.

The burden of proof is on the one making the claim.  If you can't meet that burden--and given your attitude thus far upon being requested to meet it, I assume you can't--then you got nothing.  Just drop it already! Ugh...

**Comment by kangaroo** [comment]
Who needs key management?  We can generate a random value with System.Random (Its a quality source of randomness.  You know this because its called System.Random).  Then we can xor every string with the value.  In fact, to be safe, lets xor with the value twice!  Double secure!

**Comment by ayende** [comment]
And decrypt how?
For fun, the log is from 3 months ago and rhe machine hace been repaced
twice

**Comment by masonwheeler** [comment]
@ayende You may need to recalibrate your sarcasm detector.  Consider the effect of XORing something twice with the same key "for double security."

**Comment by advancedrei** [comment]
There is no need for the complexity being suggested. This is a classic case of engineers over-engineering an unnecessary solution while missing the obvious one.

There are two types of exception messages: Informative ones that help the developer make software better. Then there are "f*** you" errors. They do nothing to help anyone. Need an example of another "f*** you" error in a Microsoft product?

![image](https://cloud.githubusercontent.com/assets/1657085/9566185/5d82c386-4ec8-11e5-9529-3652768bbf90.png)
Yeah, real helpful. There are plenty more of these types of exceptions in .NET, not just this one.

How are you, as the developer, ever supposed to determine, in production, what user input caused the problem in the first place? What if you hit on some weird situation that is hard to reproduce, and that user leaves the app and never comes back? Could be days or weeks before it comes up again, and you'll *still* be no better off.

And how are Application Insights or New Relic supposed to know, for that one exception, that it needs to grab more info? But not the other ones. That doesn't make any sense.

Anyone who says "oh you can just hook up a debugger" has never deployed a complex system into production to be used by thousands of users an hour. Hook up a debugger to a website with this problem, and every request slows to a crawl. Not an option.

My company has wasted dozens of hours tracking down these types of issues due to "f*** you" exceptions. Now multiply that by every company using the .NET Framework. It is a very real and expensive problem with an insanely simple solution.

Why is one dissenter driving the discussion? @OtherCrashOverride why don't you write your own security framework to sanitize exception messages, if you are that worried about it? You have yet to demonstrate how it would ever actually be a threat vector, and **THE** ASP.NET security guy at Microsoft said he's cool with it. Being responsible for the security of every app running on ASP.NET (and Microsoft's reputation, etc)... why does that not definitively answer the question?

The thread was closed. The security expert weighed in. The PR compiles and should be accepted. I think discussion should be locked, and everyone here move on to more pressing issues.

**Comment by OtherCrashOverride** [comment]
Here is an example of what has been discussed.  Its not a pull request. Its just real code in place of psuedo-code.  Its meant to help convey a concept, not be copy/pasted into the repo:

```
    public abstract class SecureException : Exception
    {
        public string PrivateMessage
        {
            get
            {
                if (PAL_GetDetailedExceptionInfoFeatureEnabledFlag())
                {
                    return GetPrivateMessage();
                }
                else
                {
                    return "Feature not enabled.";
                }
            }
        }



        protected SecureException(string publicMessage)
            : base(publicMessage)
        {
        }


        protected abstract string GetPrivateMessage();


        static bool PAL_GetDetailedExceptionInfoFeatureEnabledFlag()
        {
            // TODO: Read feature flag from runtime
            return true;
        }
    }


    public class DuplicateKeyException : SecureException
    {
        const string EXCEPTION_MESSAGE = "An item with the same key has already been added.";


        object key;

        public object Key
        {
            // Note: The concern here is serialization.
            // The goal is to prevent accidental leakage and
            // serialization implies intent so this value should
            // not be occluded.
            get { return key; }
        }


        public DuplicateKeyException(object key)
            : base(EXCEPTION_MESSAGE)
        {
            this.key = key;
        }


        protected override string GetPrivateMessage()
        {
            // Note: ArgumentNullException is throw instead if a key is null

            string result;

            if (key == null)
            {
                result = string.Empty;
            }
            else
            {
                if (key.GetType().IsValueType)
                {
                    result = string.Format("The key '{0}' is duplicated", key.ToString());
                }
                else
                {
                    result = string.Format("The key is an object of type {0}.", key.GetType().ToString());
                }
            }

            return result;
        }
    }


    public class FileNotFoundException : SecureException
    {
        const string EXCEPTION_MESSAGE = "Unable to find the specified file.";


        string fileName;

        public string FileName
        {
            get { return fileName; }
        }


        public FileNotFoundException(string fileName)
            : base (EXCEPTION_MESSAGE)
        {
        }


        protected override string GetPrivateMessage()
        {
            return string.Format("Could not find file '{0}'", fileName);
        }
    }
```

An interesting item presented itself here:  This model moves exceptions from abstract messages to more formal object definitions.  This means better information and a better developer experience since there is less test to be written for an exception and no need to parse it.

**Comment by OtherCrashOverride** [comment]
> The thread was closed. The security expert weighed in. The PR compiles and should be accepted. I think discussion should be locked, and everyone here move on to more pressing issues.

![It compiles! Shipt it!](http://image.slidesharecdn.com/continuous-delivery-jaxlondon-2011-johnstevenson-120314080522-phpapp01/95/continuous-delivery-jax-london-2011-63-728.jpg?cb=1331713712)

**Comment by poke** [comment]
@OtherCrashOverride [Separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns): The purpose of an exception object is to contain information about the cause of an exception; as much as information as possible in order to understand the error without having to run the debugger (which, as others explained isn’t feasible in many situations).

All your arguments are based on the idea that having more information in the exception could possibly expose security details through the log. I absolutely understand the problem you have there; but it’s not the job of the exception object to make sure that this possibly dangerous information isn’t leaked into the log: It’s the logger’s job. And it’s quite simple to do that too:

    try
    {
        CodeThatCouldThrowArgumentExceptionWithDangerousDetails();
    }
    catch (ArgumentException ex)
    {
        Log("Oh, the code failed with an ArgumentException. I know which key is the cause, but I won’t tell you.");
        EncryptedLog("The original exception was", ex);
    }

Want to encrypt the exception messages? Sure, just do that; but do it in the logger if you really need to restrict access to the information (although you could easily do that with log levels, or just file-level access restrictions to the log file).

My point is: Exception objects are supposed to be dead simple; they are often used internally without the information being read by humans or written into log files. As such, they should not contain much logic—especially not some encryption mechanism.

Sure, changing the whole exception hierarchy in a way that allows for more granular information discovery would be nice. But it would also require to change the whole exception hierarchy. And that would not only break every existing exception code, but also require to change most of the whole framework. And if you are concerned by disasters from a “compile then ship” mentality, then you should be even more concerned about a “rewrite everything” mentality.

In any way, I don’t really understand your agenda here. The issue is resolved, the team has decided on a solution, and [*that solution is already in place*](https://github.com/dotnet/corefx/issues/1187#issuecomment-135981314).

**Comment by masonwheeler** [comment]
@poke: His agenda appears to be [security theater:](https://en.wikipedia.org/wiki/Security_theater) doing something highly visible that *looks like* you're improving security to an ignorant third party, but doesn't actually make anything any more secure and mostly just manages to inconvenience everyone who has to interact with it.

**Comment by OtherCrashOverride** [comment]
> In any way, I don’t really understand your agenda here.

@poke the agenda is to make CoreCLR **useful** to a broader audience.  What developers do in their own code is of no concern to me.  However, since we all have to share the CoreCLR and CoreFX, a larger burden is placed on it: we all drink from the same well.

The exception example I posted shows the extent of what is being requested.  It does not contain encryption. Instead it allows the flexibility necessary for developers to implement encryption if they so desire.  It additionally offers other benefits to tools and testers by showing an exception object model that is diagnostically more useful than a simple text string.  There is no requirement for developers or 3rd party code to implement ``SecureException``.  It is only necessary for CoreCLR/CoreFX that we all share.

**Comment by OtherCrashOverride** [comment]
> the agenda is to make CoreCLR useful to a broader audience.

Let me clarify that the meaning of that: for some developers, it is currently either necessary to fork CoreCLR or use a different product.

**Comment by jasonwilliams200OK** [comment]
> for some developers,

You are the only one fall in that category pal. Get a life and chill the f-up already!

**Comment by OtherCrashOverride** [comment]
> You are the only one fall in that category pal.

Im not your pal, buddy.

**Comment by OtherCrashOverride** [comment]
An additional thought is to also modify ```System.Exception```:
```
public partial class Exception
{
    public virtual string PrivateMessage
    {
        get { return "No additional information provided."; }
    }
}
```

With a corresponding override in ``SecureException``, this would eliminate the need to cast to get the additional information.

**Comment by davkean** [comment]
This thread has gotten a little off the original topic. @OtherCrashOverride, feel free to open a new issue with a proposal that covers your ideas. Follow the [API Review Process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) and make sure you clearly articulate the problem you are trying to solve.
